doc/keynames/README.md

```diff
@@ -0,0 +1,477 @@
+
+- It is an empty string.
+- The last character before the invisible null terminator is a backslash `\`.
+  This called a "dangling escape", since we would expect another character after the escape character `\`.
+- It contains a Namespace (i.e. the Namespace is not "Cascading"), but the Namespace Separator `:` is not followed by a `/`.
+  (This mainly applies to Root Keys.)
+- It contains a Namespace Separator `:`, but the substring before the first `:` is not one of: `meta`, `spec`, `proc`, `dir`, `user`, `system` and `default`.
+- One of the Key Name Parts starts with an unescaped `@`.
+  For example `/elektra/@version` is an invalid Escaped Key Name.
+  Only `/elektra/\@version` would be valid.
+- It contains an Illegal Escape Sequence (see below).
+- It contains an Invalid Array Part (see below).
+- Going from left to right one Key Name Part at a time, at any point, the number of Key Name Parts that are `..` is bigger than the number of other Key Name Parts.
+  In Non-Canonical Key Names, the `.` Key Name Parts contribute to the number of other Key Name Parts.
+
+An Escape Sequence is illegal, if (one of):
+
+- It appears at the start of a Key Name Part and the escaped character is not one of: `/`, `\`, `.`, `#`, `%` or `@`.
+- It appears anywhere else and the escaped character is not one of: `/` or `\`.
+
+An Array Part is valid, if (one of):
+
+- The whole Key Name Part is `#` or `#0`.
+- The Key Name Part is `#` followed by 1-19 digits, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^64 - 1`).
+- The Key Name Part is `#` followed by `n` underscores, followed by `n+1` digits, where `n <= 18`, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^64 - 1`).
+
+Otherwise, Array Parts (i.e. Key Name Parts starting with `#`) are invalid.
+
+> _Note:_ The C-API does not allow you to construct a `Key` with an Invalid Key Name; for example `keyNew` (and `keyVNew`) will return `NULL`.
+
+## 4.1. Special Key Names
+
+Apart from Invalid Key Names, which cannot be constructed via the C-API, there are also _Reserved Key Names_.
+These can be used with the C-API (`keyNew` returns a valid `Key *`), but there might be situations, in which `Key`s with such Key Names are treated differently.
+
+A simple example is `/somewhere/_/below`.
+If used with the globbing functionality, the Key Name Part `_` will be interpreted as "match anything that is not an array element".
+
+If any of the following is true, a Key Name is a _Reserved Key Name_:
+
+- A Key Name Part starts with `_`.
+- A Key Name Part starts with `@`.
+- A Key Name Part starts with `%` and the Key Name Part is more than one byte long.
+
+<!-- TODO (kodebach): Is keyIsInactive still relevant? -->
+
+There are also _Inactive Key Names_ (also _Hidden Key Names_).
+These are Key Names with a Key Name Part that starts with a `.`.
```

Comment:
@markus2330 @mpranj As the TODO says, is `keyIsInactive` still relevant or should it be removed any way?

Reply:
Irrelevant, please remove.